src/states/node.rs

```diff
@@ -722,43 +728,81 @@ impl Node {
     // Acknowledge reception of the message and broadcast to our section if necessary
     // The function is only called when we are in the destination authority
     fn ack_and_broadcast(&mut self,
-                         signed_msg: &SignedMessage,
+                         signed_msg: SignedMessage,
                          route: u8,
                          hop_name: XorName,
                          sent_to: &BTreeSet<XorName>) {
         self.send_ack(signed_msg.routing_message(), route);
         // If the destination is our section we need to forward it to the rest of the section
         if signed_msg.routing_message().dst.is_multiple() {
-            if let Err(error) = self.send_signed_message(signed_msg, route, &hop_name, sent_to) {
+            // TODO: maybe eventually don't clone signed_msg just so we can log it on error?
+            if let Err(error) =
+                self.send_signed_message(signed_msg.clone(), route, &hop_name, sent_to) {
                 debug!("{:?} Failed to send {:?}: {:?}", self, signed_msg, error);
             }
         }
     }
 
+    // Verify the message, then, if it is for us, handle the enclosed routing message; if not,
+    // forward it.
+    //
+    // If we are the originator or accumulator of the message, then `hop_name` is our name;
+    // otherwise, it is the name of the node which forwarded us the message.
     fn handle_signed_message(&mut self,
-                             signed_msg: SignedMessage,
+                             mut signed_msg: SignedMessage,
                              route: u8,
                              hop_name: XorName,
                              sent_to: &BTreeSet<XorName>)
                              -> Result<(), RoutingError> {
-        signed_msg.check_integrity(self.min_section_size())?;
+        let hop_prefix = self.peer_mgr
+            .routing_table()
+            .find_section_prefix(&hop_name)
+            .ok_or(RoutingTableError::NoSuchPeer)?;
+        let section_list = if signed_msg.routing_message().src.is_client() ||
+                              self.in_authority(&signed_msg.routing_message().dst) {
+            // TODO: we should still verify the sending nodes when the sender is our own section
+            // or group, but doing so causes a lot of verification failures at the moment.
+            // (In any case don't call add_relaying_section if from our own section.)
```

Comment:
I saw similar issues, and I think some of those failures might be because as a candidate, the second node expects two signatures in its `NodeApproval`, because it already considers itself part of the section. A simple workaround is to just exclude messages that were sent to us as a client here, i.e. if `dst` is a `Client` with our key. Does that resolve the failures?

Also, I think it might be a good idea to include our own section's list if we are the first node in our section to receive this message, i.e. if the `hop_name` is not in our section:
We need to ensure that either all nor none of the nodes in the section accept the message, and adding our own section list would reduce that problem to all nodes in our section accepting our own list of our own section, as opposed to our list of a neighbouring section. But that probably needs some more discussion.